### ISSUE TITLE
"Undeclared name" errors for accessible identifiers in same package (after go1.3.3)

### ISSUE THREAD

**Issue by zrhmn** [microsoft/vscode-go#2858]
**Describe the bug**  
After I recently updated to go1.3.3, I get a lot of errors about vars, funcs etc. being undefined. These vars are defined in different files, but in the same package, but gopls/vscode-go (I can't say which) is unable to find a reference to them.

The gopls log doesn't show any errors, but if I hover over the squiggled name, or ⌘+Click them, an error shows up in gopls log like:

```
[Error - 8:10:31 AM] Request textDocument/definition failed.
  Message: no object for ident Exported
  Code: 0
``` 

**VSCode settings**  
`Preferences: Open Settings (JSON)`
```json
{
    "editor.cursorBlinking": "phase",
    "editor.fontFamily": "Go Mono",
    "editor.renderControlCharacters": true,
    "editor.renderLineHighlight": "all",
    "editor.renderWhitespace": "boundary",
    "editor.rulers": [80],
    "editor.smoothScrolling": true,
    "editor.tabSize": 2,
    "explorer.confirmDelete": false,
    "explorer.confirmDragAndDrop": false,
    "explorer.openEditors.visible": 0,
    "files.insertFinalNewline": true,
    "telemetry.enableTelemetry": false,
    "workbench.startupEditor": "none",
    "clang.cflags": ["-std=c99", "-Wall", "-Wextra"],
    "javascript.updateImportsOnFileMove.enabled": "always",
    "tslint.autoFixOnSave": true,
    "tslint.jsEnable": true,
    "window.zoomLevel": 0,
    "markdown.preview.fontSize": 16,
    "editor.fontLigatures": true,
    "workbench.colorTheme": "Monokai Dimmed",
    "breadcrumbs.enabled": true,
    "go.useLanguageServer": true,
    "go.testFlags": [
        "-v"
    ],
    "cmake.configureOnOpen": true,
    "editor.cursorSmoothCaretAnimation": true,
    "workbench.iconTheme": "vscode-icons",
    "scm.defaultViewMode": "tree"
}
```

**To Reproduce**  
1. Open a go module source in VSCode.
2. Create a global variable in one file.
3. Access that variable in another file.

**Expected behavior**  
To have VSCode not report the variables defined and accessible within the package as undefined, and "Go to definition" feature working.

**Screenshots**  
![image](https://user-images.githubusercontent.com/4042217/67327247-a38b1500-f4e5-11e9-9f83-1f406245b85b.png)

**System information:**  
 - OS: Mac OSX
 - Extension Version: 0.11.7
 - VS Code Version: 1.39.2

**Additional context**
None.

**Comment by stamblerre** [comment]
Can you share the output of `gopls -rpc.trace -v check path/to/file.go`?

**Comment by zrhmn** [comment]
```
2019/10/22 16:05:13 Info:2019/10/22 16:05:13 Build info
----------
golang.org/x/tools/gopls v0.1.7
    golang.org/x/tools/gopls@v0.1.7 h1:YwKf8t9h69++qCtVmc2q6fVuetFXmmu9LKoPMYLZid4=
    golang.org/x/sync@v0.0.0-20190423024810-112230192c58 h1:8gQV6CLnAEikrhgkHFbMAEhagSSnXWGV915qUMm9mrU=
    golang.org/x/tools@v0.0.0-20190918214516-5a1a30219888 h1:ER45Jz0UDQ3e6em1lwXVwuPf96lvyQogb7m+gEbsoPg=
    golang.org/x/xerrors@v0.0.0-20190717185122-a985d3407aa7 h1:9zdDQZ7Thm29KFXgAX/+yaf3eVbP7djjWp/dXAppNCc=

Go info
-------
go version go1.13.3 darwin/amd64

GO111MODULE=""
GOARCH="amd64"
GOBIN="/Users/fziaur094/.local/bin"
GOCACHE="/Users/fziaur094/Library/Caches/go-build"
GOENV="/Users/fziaur094/Library/Application Support/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GONOPROXY="github.comcast.com"
GONOSUMDB="github.comcast.com"
GOOS="darwin"
GOPATH="/Users/fziaur094/.local/lib/go"
GOPRIVATE="github.comcast.com"
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/Cellar/go/1.13.3/libexec"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/Cellar/go/1.13.3/libexec/pkg/tool/darwin_amd64"
GCCGO="gccgo"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/Users/fziaur094/Workspace/@sandbox/go/gopls-check/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/58/t6vp6v297v17qnyjz2662_fr0000gp/T/go-build509819033=/tmp/go-build -gno-record-gcc-switches -fno-common"
2019/10/22 16:05:13 Info:2019/10/22 16:05:13 19.757964ms for GOROOT= GOPATH=/Users/fziaur094/.local/lib/go GO111MODULE= PWD=/Users/fziaur094/Workspace/@sandbox/go/gopls-check go "env" "GOMOD", stderr: <<>>
2019/10/22 16:05:13 Info:2019/10/22 16:05:13 29.287783ms for GOROOT= GOPATH=/Users/fziaur094/.local/lib/go GO111MODULE= PWD=/Users/fziaur094/Workspace/@sandbox/go/gopls-check go "list" "-e" "-json" "-compiled=true" "-test=true" "-export=false" "-deps=true" "-find=false" "--" "builtin", stderr: <<>>
2019/10/22 16:05:13 Info:2019/10/22 16:05:13 29.265414ms for GOROOT= GOPATH=/Users/fziaur094/.local/lib/go GO111MODULE= PWD=/Users/fziaur094/Workspace/@sandbox/go/gopls-check go "list" "-m" "-json" "all", stderr: <<>>
2019/10/22 16:05:13 Info:2019/10/22 16:05:13 16.035754ms for GOROOT= GOPATH=/Users/fziaur094/.local/lib/go GO111MODULE= PWD=/Users/fziaur094/Workspace/@sandbox/go/gopls-check go "env" "GOMOD", stderr: <<>>
2019/10/22 16:05:13 Info:2019/10/22 16:05:13 18.961124ms for GOROOT= GOPATH=/Users/fziaur094/.local/lib/go GO111MODULE= PWD=/Users/fziaur094/Workspace/@sandbox/go/gopls-check go "list" "-m" "-json" "all", stderr: <<>>
2019/10/22 16:05:13 Info:2019/10/22 16:05:13 15.275052ms for GOROOT= GOPATH=/Users/fziaur094/.local/lib/go GO111MODULE= PWD=/Users/fziaur094/Workspace/@sandbox/go/gopls-check go "list" "-e" "-json" "-compiled=true" "-test=true" "-export=false" "-deps=true" "-find=false" "--" "/Users/fziaur094/Workspace/@sandbox/go/gopls-check", stderr: <<>>
2019/10/22 16:05:14 Info:2019/10/22 16:05:14 251.939458ms for GOROOT= GOPATH=/Users/fziaur094/.local/lib/go GO111MODULE= PWD=/Users/fziaur094/Workspace/@sandbox/go/gopls-check go "list" "-e" "-json" "-compiled=true" "-test=true" "-export=false" "-deps=true" "-find=false" "--" "/Users/fziaur094/Workspace/@sandbox/go/gopls-check/check.go", stderr: <<>>
2019/10/22 16:05:14 Info:2019/10/22 16:05:14 go/packages.Load
	packages = 1
2019/10/22 16:05:14 Info:2019/10/22 16:05:14 go/packages.Load
	package = command-line-arguments
	files = [/Users/fziaur094/Workspace/@sandbox/go/gopls-check/check.go]
```

**Comment by stamblerre** [comment]
Can you run `go list -e -json -compiled=true -test=true -deps=true /Users/fziaur094/Workspace/@sandbox/go/gopls-check` from inside the `/Users/fziaur094/Workspace/@sandbox/go/gopls-check` directory?

**Comment by zrhmn** [comment]
```{
	"ImportPath": "/Users/fziaur094/Workspace@sandbox/go/gopls-check",
	"Match": [
		"/Users/fziaur094/Workspace@sandbox/go/gopls-check"
	],
	"Error": {
		"ImportStack": [
			"/Users/fziaur094/Workspace@sandbox/go/gopls-check"
		],
		"Pos": "",
		"Err": "can only use path@version syntax with 'go get'"
	}
}```

**Comment by jayconrod** [comment]
@zrhmn Thanks for investigating. This seems to be a regression in go1.13.3. I've opened golang/go#35115 to track it.

`@` is not allowed in module paths or import paths, but it should be allowed in directory and file names.

**Comment by floodcode** [comment]
Having same issues on vscode 1.39.2 + go extension 0.11.7 + golang 1.13.1. And i have no `@` in path.

**Comment by stamblerre** [comment]
@floodcode: If you are using `gopls`, can you open a new issue on the [Go issue tracker](https://github.com/golang/go/issues/new) with the information requested in the [Troubleshooting guide](https://github.com/golang/tools/blob/master/gopls/doc/troubleshooting.md)?

**Comment by floodcode** [comment]
@stamblerre Sure

**Comment by floodcode** [comment]
Investigated a little bit deeper. It also reproduces in go v1.13.3. But once i initialize go module in my project (via `go mod `), it creates `go.mod` file and problem disappears.

**Comment by ramya-rao-a** [comment]
Since it has been established that this is due to behavior changes in Go, am closing this issue in favor of the upstream issues as there is no action item on the extension's side at the moment

**Comment by zrhmn** [comment]
_(Posted this reply earlier on another issue by mistake)._

For other people following this issue, the problem persists even with go1.13.4.

The fix was not backported to 1.13 because [the cherrypick was rejected](https://github.com/golang/go/issues/35119). I would have preferred to be a part of that discussion, but it's my own fault for not following the maze of issue-links. Too late now. 🤷‍♂

The fix is [here](https://go-review.googlesource.com/c/go/+/202763/1/src/cmd/go/internal/search/search.go#368) for anyone interested. May still make it in go1.14.

**Comment by GTANAdam** [comment]
I'm currently having this issue, would be great to have this annoying error not popup during development otherwise everything compiles fine anyway.. it's just annoyance.
Would be great to have the patch backported, thanks!
